bpf/lib/nat.h

```diff
@@ -1040,6 +1040,8 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
 
 	if (snat_v4_rev_nat_can_skip(target, &tuple))
 		return NAT_PUNT_TO_STACK;
+	if (egress_gw_rev_snat_ct_needed(ip4))
+		target->egress_gateway = true;
```

Comment:
I believe this won't work as-is. The reply hasn't been RevSNATed yet (that only happens in `snat_v4_rewrite_ingress()`), so `ip4->daddr` is still the EgressIP and the packet won't match any EgressGW policy entry. We also don't capture the `from_local_endpoint` condition, so EgressGW replies to local endpoints would get double-accounted.

I wonder if we can simplify the whole logic quite a bit ... when initially creating the SNAT & RevSNAT entries on Egress, just store a `needs_ct` flag into each entry (like the `host_local` flag today - or maybe we can even re-use that flag? It originated from the same purpose). And then just use that flag to decide whether a `ct_lookup()` is needed for subsequent packets.

Reply:
Looks good! Cool that we can re-use the existing flag for both cases!

One slightly annoying thing is that we duplicate the `needs_ct` decision logic in two places - `snat_v4_track_connection()` and then again in `snat_v4_new_mapping()`. It was like that already for `host_local`, so no worries. But besides the usual risk of the code drifting apart - I believe we might want to backport this to `v1.12`, and the NAT code is **very** tricky complexity-wise... 

I think we could lift that whole logic into `snat_v4_nat_handle_mapping()`, and drive both functions with it (only call `snat_v4_track_connection()` if needed, pass `needs_ct` as parameter to `snat_v4_new_mapping()`). And then just look at `state->needs_ct` in the revSNAT path. If that doesn't get *too* invasive, I think it's worth doing the change right away  (instead of cleaning up after). wdyt?

Reply:
>Also I think I should add the relevant tests here, including [yours](https://github.com/cilium/cilium/commit/414d8185dce5b750dab87d75ec9e3f5d21ea3fa3).

Absolutely! It was a quick hack, but it's good enough to increase the coverage. And we can always improve it later.

Reply:
> The reply packet doesn't match the CT entry, because revSNAT hasn't been done at that point yet =/

:exploding_head: oh joy. Yes, sure looks like it - on egress we're creating the CT_EGRESS entry with the pre-NAT tuple, but on ingress the CT lookup then uses a pre-revNAT tuple. So the CT lookup won't find a match at all, and we just create a new CT_INGRESS entry instead? wow. Are you actually seeing that additional CT entry in your testing?

So we'd basically need to revSNAT the tuple that's used for the lookup. And this is already affecting host-originating connections. Then maybe let's just go with your minimal EgressGW fix after all, and sort out the whole reply-path mess in a separate PR? :disappointed:

Ugh. And it's even more complicated. We're jumping through all sort of hoops to extract the L4 ports in `snat_v4_nat()` - but then `ct_lookup4()` of course does its own extraction again.

Reply:
>Yes, I checked that an additional entry is indeed created.

nice / ugh. At least that confirms our understanding of the code.